include/aspect/particle/particle.h

```diff
@@ -190,6 +205,23 @@ namespace aspect
         get_location () const;
 
         /**
+         * Set the location of this particle. Note that this does not check
+         * whether this is a valid location in the simulation domain.
+         *
+         * @param [in] new_loc The new location for this particle.
+         */
+        void
+        set_reference_location (const Point<dim> &new_loc);
```

Comment:
(But I don't know whether that's possible in your algorithm, haven't read that far down in the patch yet :-) )